Add easy to digest selector macros for EVP_PKEYs

levitte opened the pull request:
These are meant to be used with functions like
OSSL_ENCODER_CTX_new_by_EVP_PKEY()

The OSSL_ENCODER_CTX_new_by_EVP_PKEY() manual is also expanded on the
topics of output types and selections.

levitte commented:
I've started referring to these in another PR, without even thinking...  #13096.

These are designed to make life easier for the users, and constitute what we normally use.

kaduk commented:
Did you fail to add the file with the actual `#define`s for `EVP_PKEY_PRIVATE_KEY`/etc.?

levitte commented:
> Did you fail to add the file with the actual `#define`s for `EVP_PKEY_PRIVATE_KEY`/etc.?

Indeed I did.  Now corrected.

I also changed my mind on one of the macros.  `EVP_PKEY_PRIVATE_KEY` has been renamed to `EVP_PKEY_KEYPAIR`, to correctly reflect what we expect from an `EVP_PKEY`...  and also leave space for a future possibility that an EVP_PKEY with a private key doesn't necessarily contain a public key.  There's been enough confusion on that topic as it is (I'm sure @romen agrees :wink:)

slontis commented:
> > Did you fail to add the file with the actual `#define`s for `EVP_PKEY_PRIVATE_KEY`/etc.?
> 
> Indeed I did. Now corrected.
> 
> `EVP_PKEY_PRIVATE_KEY` has been renamed to `EVP_PKEY_KEYPAIR`, to correctly reflect what we expect from an `EVP_PKEY`.

I think the future might be right now. The ec app has this nugget in it..
EC_KEY_set_enc_flags(eckey, EC_PKEY_NO_PUBKEY);

levitte commented:
Yeah, and I think we have determined that EC is an odd exception to the (documented) rule, see the discussion in #12612 and https://mta.openssl.org/pipermail/openssl-project/2020-October/002290.html

slontis commented on doc/man3/OSSL_ENCODER_CTX_new_by_EVP_PKEY.pod:
```diff
@@ -39,7 +39,11 @@ OSSL_ENCODER_CTX_set_passphrase_ui
+L<OSSL_ENCODER_to_bio(3)>.  I<output_type> determines the final output
+encoding, and I<selection> can be used to select what parts of the I<pkey>
+should be included in the output.  I<output_type> is further discussed in
+L</Output types> below, and I<selection> is further describe in
```
described

levitte replied:
Fixed

slontis commented:
I am still wondering if this is naming scheme is painting you into a corner.
The logic behind renaming to keypair so that we could maybe use private to just mean private, is then a bit weird because public doesnt just mean public.

keypair is actually 'all' or 'private_with_dependancies'.

I guess we could also have these??
private_only
public_only

levitte commented:
> I am still wondering if this is naming scheme is painting you into a corner.
> The logic behind renaming to keypair so that we could maybe use private to just mean private, is then a bit weird because public doesnt just mean public.

... and KEYPAIR doesn't necessarily mean just the keypair.  Both include key parameters if that key depends on it.

levitte commented:
Ping!

Note that this is extensible.  The ones defined in this PR are the ones *currently* used, and that should be sufficient, no?

slontis commented:
> Do you currently have a purpose for them?

The EC case is not just going to disappear because you want it to :)
I feel we are going in circles here. Extensible implies that you have considered that it might be extended,
and in that case - can you tell me what names you would use for 'just a private key' and 'just a public key'.

levitte commented:
@slontis, I think you misunderstand the point of "easy to digest", and the imprecision implied.  The manual does point it out:

> These are only indications, the encoder implementations are free to
> determine what makes sense to include in the output, and this may depend on
> the desired output.  For example, an EC key in a PKCS#8 structure doesn't
> usually include the public key.

So, for an encoding of a private key to, for example, PKCS#8, the selector macro to use would be `EVP_PKEY_KEYPAIR`...  the encoders will simply do what they do best.  The point is also that the user doesn't have to know a precise set of bits to be used with each key type.

But to answer your question, there are two possible responses:

-   use the `OSSL_KEYMGMT_SELECT_` macros if you want more precision.  Do notice that the encoders and decoders will not necessarily "obey" more precisely, you'll still get parameters in your PKCS#8 output even if you only specified `OSSL_KEYMGMT_SELECT_PRIVATE_KEY`, when the standards say they must be there.
-   For private key only, it would be easy to add `EVP_PKEY_PRIVATE_KEY`.

slontis commented:
Are you supposed to be in a meeting right now..

levitte commented:
Something to keep in mind is that the macros suggested here are meant to be used together with an EVP_PKEY, which is a bit of a "whatever" type from a user perspective, all the users know is that they may contain key parameters, or a public key, or a private key, with whatever entourage they require (for private and public keys, the "entourage" would be key parameters, and for private keys, that most often implies a public key as well).

openssl-machine commented:
This pull request is ready to merge

levitte commented:
Merged

140eee2b3b3090107aea2d09ab10eb737fb1e5bb Add easy to digest selector macros for EVP_PKEYs
